event.cpp

```diff
@@ -38,21 +38,30 @@ TCHAR *message_string(unsigned long error) {
 /* Log a message to the Event Log */
 void log_event(unsigned short type, unsigned long id, ...) {
   va_list arg;
-  int count;
-  TCHAR *s;
   TCHAR *strings[NSSM_NUM_EVENT_STRINGS];
+  TCHAR *buffer = NULL;
+  int len;
+
+  TCHAR *format = message_string(id);
+  if (!format) return;
 
   /* Open event log */
   HANDLE handle = RegisterEventSource(0, NSSM_SOURCE);
   if (! handle) return;
 
   /* Log it */
-  count = 0;
   va_start(arg, id);
-  while ((s = va_arg(arg, TCHAR *)) && count < NSSM_NUM_EVENT_STRINGS - 1) strings[count++] = s;
-  strings[count] = 0;
+  len = _vsctprintf(format, arg) + 1;
+  buffer = (TCHAR*)malloc(len * sizeof(TCHAR));

```

Comment:
`len` should be checked against `NSSM_ERROR_BUFSIZE` or a defined max value.

I would also prefer to allocate buffer statically `TCHAR buffer[NSSM_ERROR_BUFSIZE]` instead malloc/free - but I don't have a strong argument for this